### ISSUE TITLE
Multiselect via ctrl+d case insensitivity makes it useless

### ISSUE THREAD

**Issue by ds0nt** [microsoft/vscode#18239]
- VSCode Version: Code 1.8.1 (ee428b0eead68bf0fb99ab5fdc4439be227b6281, 2016-12-19T14:41:20.664Z)
- OS Version: Linux x64 4.4.38-1-MANJARO
- Extensions:

|Extension|Author|Version|
|---|---|---|
|Go|lukehoban|0.6.51|

---

Steps to Reproduce:

consider this block of code:
```
	accountQryConn   *grpc.ClientConn
	accountQryClient proto.AccountQueryClient
```

when I select 'account' and hit alt+d a few times it will select it like this.

![2017-01-06-150041_626x115_scrot](https://cloud.githubusercontent.com/assets/1015732/21731141/eb1079a8-d420-11e6-89a9-280b0f8d463b.png)

this is fine, but when I do this, I am doing it to replace it with something else. So I'll start typing:

![2017-01-06-150136_731x106_scrot](https://cloud.githubusercontent.com/assets/1015732/21731163/09140b68-d421-11e6-913e-a0bf92e9ed67.png)

but now my case is broken for my second string.

This is fine in a small example, but very annoying. 

Currently the case matches whatever is highlighted in the search widget. Behold what happens if I turn off case sensitivity: 

![2017-01-06-150329_1605x153_scrot](https://cloud.githubusercontent.com/assets/1015732/21731213/4eb44f16-d421-11e6-8fba-8f90ed207420.png)

This is incredibly annoying. I can't think of a time when you would want to multiselect case-insensitively with ctrl+d. It simply selects the wrong stuff, because in code it's always case sensitive. So depending on whatever I did last with ctrl+f, my ctrl+d either works or is annoying.

I think ctrl+d should always be case sensitive, or have a button that's not search-tied that does a case sensitive string selection match across the file

My two cents.

**Comment by ds0nt** [comment]
These are my atom habits, btw. :)

**Comment by ds0nt** [comment]
but vscode has good golang support so I'm giving it a chance.

**Comment by ds0nt** [comment]
I'd prefer it if I could select all the instances of 'account' one by one, and then type something else to replace it.. like 'voucher'. Following which, I'll select the "Account" and replace it as well with "Voucher". This makes sense to me, as Voucher != voucher in terms of what a compiler thinks

**Comment by sirgallifrey** [comment]
\+ 1 to this request.
My case is similar, but I make a selection and the vscode makes a case insensitive highlight, so I can press ctrl + shit + L and select all de occurrences (case insensitive). Then I have, lets say `RateField` and `ratefield` selected, and those are not the same thing, which makes this feat less useful.
I was thinking as this function is related to the "editor.selectionHighlight" we could have a "editor.selectionHighlightCaseMatch" to choose from sensitive or insensitive, in that fashion the default behavior won't change.

**Comment by nicolasg-da** [comment]
I have the same issue in Haskell, where `value` and `Value` are two very different things.

Maybe it would be possible to set a language specific option for this?
So keep the current behaviour but add an option to overwrite it for certain languages where casing matters?

**Comment by leocaseiro** [comment]
Try `Alt+C`(pc) or `Option+Command+C`(mac) to `toogleFindCaseSensitive`

**Comment by sirgallifrey** [comment]
Thank you so much!! 
That is the most obscure config ever! I would never find it on my own... Thanks again, you made my vscode so mush better!


![272185fa61322668c896868518f3a8ef_feels-good-man-meme-feels-good-man_430-397](https://cloud.githubusercontent.com/assets/2331645/24908987/955737e6-1e98-11e7-8ee4-f627d20d4c2e.jpeg)

**Comment by michaelswe** [comment]
Please make findCaseSensitive a setting so we can have case sensitivity by default. Everytime VSCode is restarted the case sensitivity toggles back to the default, ie. not case sensitive.

This has bitten me many times when using cmd-d for multiselecting in "refactoring sessions"

**Comment by altla** [comment]
findCaseSensitive *really* needs to be a setting. Without case sensitivity, "next match" is pretty much useless.

**Comment by ViggoV** [comment]
+1! I very much agree! This bugs me every day!

**Comment by leocaseiro** [comment]
GUYSSSSS!

I'm so happy, so so happy!

The new release [1.13](https://code.visualstudio.com/updates/v1_13#_preserving-view-state-for-resource-editors) now preserves the settings of `toogleFindCaseSensitive` state.

👍  🍻  🎉  🎈  ❤️  😃  🍷  👍  🍻  🎉  🎈  ❤️  😃  🍷  👍  🍻  🎉  🎈  ❤️  😃  🍷

**Comment by msageryd** [comment]
Great news!
This was the first thing I looked for in the change log. Must have missed it. Thank you for letting us know.

**Comment by rebornix** [comment]
@leocaseiro @msageryd thanks for calling it out. I added this for half a month in Insider but forgot to mention it in release note :)

**Comment by altla** [comment]
@rebornix do you have a link to that commit?

**Comment by alexandrudima** [comment]
PR #36682 (for #8869) makes it that when starting with a collapsed selection, Command+D/Control+D will now search with `wholeWords: true and caseSensitive: true`.

Command+D/Control+D will still respect the find widget settings when starting with a non-collapsed selection.

**Comment by ViggoV** [comment]
@alexandrudima I'm not sure what you mean by collapsed selection? I don't see why the cmd+d/ctrl+d selection can't just have its own, permanent settings. There is absolutely no scenario where I would do a case insensitive multiselect and I would personally like to keep wholeWords set to false (eg. if I copy multiple vars with the same prefix and want to replace that)..

**Comment by alexandrudima** [comment]
@ViggoV I see your point. But the problem is that what works well for one person will not necessarily work well for a large set of other people. Not everyone would like to have permanent settings for Ctrl-D.

This is what I call a collapsed selection:
![image](https://user-images.githubusercontent.com/5047891/31994815-4cf6c05e-b982-11e7-8e70-782bf6c610c8.png)

This is what I call a non-collapsed selection:
![image](https://user-images.githubusercontent.com/5047891/31994829-614530f4-b982-11e7-9041-672ce6c52ff4.png)

Please try out the latest release in the Insiders Channel and provide feedback after you have tried it out.

**Comment by JakeKantzer** [comment]
I agree with @ViggoV here - a case-insensitive Ctrl+D is literally useless to me. At least the option of setting its case sensitivity would be good. I can't think of another text editor that handles this the way Code does; both Atom and Sublime are case sensitive for this operation and with good reason. The new behavior is better but still isn't great.

That said, I think [my extension](https://github.com/JakeKantzer/vscode-case-sensitive-add-next) is a pretty good bandaid. The highlighting doesn't match because it still shows the default Ctrl+D behavior but otherwise I tried to make the command my extension adds work like you would want this functionality to behave. I have my editor set up so my extension's action overrides Ctrl+D and the default is Ctrl+Shift+D.

**Comment by ViggoV** [comment]
Thanks @JakeKantzer! Works like a charm.. As for the question on whether multiselect should be case-insensitive in the first place I will stop arguing against it as soon as anyone shows me a valid and non-edge use case for it, where you would actually want to replace matches of differing case with text of the same case. The collapsed/non-collapsed solution simply doesn't cut it, since you might want to copy `{ momName: "Mommy", momAge: "Really old" }` and replace `mom` with `dad`. In my opinion that solution just puts more confusion into the mix, which may well annoy a lot of other people. But but but, I'll shut up about it as soon as someone shows me a reasonable use case to justify case-sensitivity  being the default option and the one it reverts to.. ;)

**Comment by maranomynet** [comment]
> "I think ctrl+d should always be case sensitive"

+1

...or at the very least it should NOT be tied in with the `Ctrl+F` state, as their use-cases/-patterns are normally very different.

**Comment by garajo** [comment]
@leocaseiro `toogleFindCaseSensitive` gives me 'Unknown configuration setting'. Is that the full namespace?

**Comment by leocaseiro** [comment]
Hi @garajo, I don't think there's a setting for that. It's a shortcode only.

You can also click to toggle the case sensitive while searching `Ctrl+F`(pc) or `Cmd+F`(mac):

![gthjvar730](https://user-images.githubusercontent.com/940070/40028799-2fb84bc0-5824-11e8-9ef8-3a76a628db78.gif)

PS: It works on Global Search to `Ctrl+Shift+F`(pc) or `Cmd+Shit+F`(mac).

**Comment by alvitawa** [comment]
@leocaseiro So is this toggle supposed to be preserved after exiting vscode? I see that you said last year that version 1.13 would make toogleFindCaseSensitive a setting, but when I restart vscode it forgets wether it was set to case sensitive or case insensitive.

**Comment by mjpizz** [comment]
I'm having a similar issue - it seems I have to toggle case insensitivity pretty often. Maybe it's just per-workspace? I interact with a lot of different repositories and it has surprised me every time.

Would definitely put +1 behind having my case-sensitivity toggle persisted globally, rather than having to toggle it again for each workspace.

**Comment by THoisington** [comment]
So If I understand the original issue, if I want to replace Account with Profile in Daniel's code images it is impossible to do so without them all being written in the same case ("Profile" or "profile"). 

If the argument is that it would be better if that cmd+d operation adjusted to the case of what it is replacing, I agree. But maybe I am misreading the comments here and that is already possible somehow 😅

**Comment by tj** [comment]
👍 one of the first things I noticed, a setting would be great

**Comment by greenais** [comment]
+100 for separtion of ctrl-D and ctrl-F spaces. 
Really other proven_to_be_good editors keep ctrl-D case-sensitive for the good reasons (already clearly stated abowe). Why MS should be different in a bad way?
I personnaly don't care about default setting - choose whatever you like - since I will be able to set desired behaviour for ctrl-D _separately_. 
Otherwise after each ctrl-F I _have to_ remember to fire alt-C again in order to be safe from "nice surprises".

**Comment by kanlukasz** [comment]
It is true that it should be **definitely separated.**
`toogleFindCaseSensitive` is not a solution, because when I am looking for something (ctrl + f), in most cases i would like to see all words with all case sensitives. Unlike when text is selected (ctrl+d)

**Comment by fnky** [comment]
I also vouch for a better user experience in this regard. If the editor would remember the settings that I have selected during Find, Find & Replace and Multiple Occurrences across sessions, then it would less confusing.

At the moment it defaults back to no settings being selected when you open a new editor.

**Comment by McNerdius** [comment]
Agreed, it needs to be a persisted setting.  Or a new default, or decoupled from ctrl+F, or any/all of the above.

Ctrl+D ... oops it was case insensitive ... Ctrl+U, Alt+C, start over...  Every time ?  Probably not quite, but it feels like it.  

Case insensitive as a default here is just odd/unintuitive IMO.

**Comment by pmarreck** [comment]
Chiming in to vouch for Control-D case sensitivity as a permanent option that defaults to "case-sensitive"

**Comment by tj** [comment]
👍  super annoying, the default is never what I actually want

**Comment by msageryd** [comment]
What's the status of this issue? It seemed to have been solved, but it really isn't. Or have I missed something?

As long as this setting is not a permanent setting it's of no use. I don't know when it happens, but now and then my ctrl-d gets case insensitive. Is it still connected to the ctrl-f setting? 

Oh, how great this worked back in the SublimeText days. I simply don't trust ctrl-d in VSCode, and that's a bad thing in an editor.

**Comment by kanlukasz** [comment]
>  I simply don't trust ctrl-d in VSCode,

You can change this shortcut and make it similar to Sublime Text
Or you can install https://marketplace.visualstudio.com/items?itemName=ms-vscode.sublime-keybindings

**Comment by McNerdius** [comment]
@msageryd and everyone else - in the meantime: [Case-Sensitive Add Next Occurrence](https://marketplace.visualstudio.com/items?itemName=jacobkantzer.case-sensitive-add-next) extension.  It's bound to `ctrl+shift+d` by default, but i went all in...

      {
         "key": "ctrl+d",
         "command": "-editor.action.addSelectionToNextFindMatch",
         "when": "editorFocus"
      },
      {
         "key": "ctrl+d",
         "command": "extension.caseSensitiveAddNext",
         "when": "editorFocus"
      }

**Comment by Radivarig** [comment]
There is an extension [multi-cursor-case-preserve](https://marketplace.visualstudio.com/items?itemName=Cardinal90.multi-cursor-case-preserve) which keeps capital cases while typing in replacement string.

**Comment by initplatform** [comment]
I don't want to have to add an extension.. how about a setting that just lets me specify the defaults. 

When I add new projects I always get bit by this. 

There was a good suggestion a couple years ago in another thread -> 

https://github.com/microsoft/vscode/issues/4309#issuecomment-276530055

**Comment by initplatform** [comment]
BTW.... is there even anyone lobbying to keep ctrl+d case insensitive as default? 
Was it a conscious choice to make it insensitive in the first place?

Not being difficult, seriously curious as to what advantage case insensitive ctrl+d gives?

I can't think of a case insensitive programming language, though I'm sure there may be one.

**Comment by tj** [comment]
Please :D the default doesn't make much sense in a programming context
